src/layout/CustomButton/CustomButtonComponent.tsx

```diff
@@ -0,0 +1,202 @@
+ */
+const isClientAction = (action: CBTypes.CustomAction): action is CBTypes.ClientAction => action.type === 'ClientAction';
+/**
+ * A type guard to check if the action is an action that requires a server side call
+ */
+const isServerAction = (action: CBTypes.CustomAction): action is CBTypes.ServerAction => action.type === 'ServerAction';
+
+function useHandleClientActions(): UseHandleClientActions {
+  const dispatch = useAppDispatch();
+  const currentDataModelGuid = useCurrentDataModelGuid();
+  const { navigateToPage, navigateToNextPage, navigateToPreviousPage } = useNavigatePage();
+
+  const frontendActions: ClientActionHandlers = {
+    nextPage: promisify(navigateToNextPage),
+    previousPage: promisify(navigateToPreviousPage),
+    navigateToPage: promisify<ClientActionHandlers['navigateToPage']>(async ({ page }) => navigateToPage(page)),
+  };
+
+  const handleClientAction = async (action: CBTypes.ClientAction) => {
+    if (isSpecificClientAction('navigateToPage', action)) {
+      return await frontendActions[action.name](action.metadata);
+    }
+    await frontendActions[action.name]();
+  };
+
+  return {
+    handleClientActions: async (actions) => {
+      for (const action of actions) {
+        await handleClientAction(action);
+      }
+    },
+    handleDataModelUpdate: async (updatedDataModels) => {
+      const currentDataModelUpdates = currentDataModelGuid && updatedDataModels[currentDataModelGuid];
+      if (currentDataModelUpdates) {
+        dispatch(FormDataActions.fetchFulfilled({ formData: flattenObject(currentDataModelUpdates) }));
+      }
+    },
+  };
+}
+
+type PerformActionMutationProps = {
+  action: CBTypes.CustomAction;
+  buttonId: string;
+};
+
+type UsePerformActionMutation = {
+  mutation: UseMutationResult<ActionResult>;
+  performAction: (props: PerformActionMutationProps) => Promise<void>;
+};
+
+type PerformActionMutationError = {
+  response: {
+    data: {
+      error: {
+        message: string;
+        code: string;
+        metadata: Record<string, unknown>;
+      };
+    };
+  };
+};
+
+function usePerformActionMutation(): UsePerformActionMutation {
+  const { doPerformAction } = useAppMutations();
+  const { partyId, instanceGuid } = useNavigationParams();
+
+  const mutation = useMutation({
+    mutationFn: async ({ action, buttonId }: PerformActionMutationProps) => {
+      if (!instanceGuid || !partyId) {
+        throw Error('Cannot perform action without partyId and instanceGuid');
+      }
+      return doPerformAction.call(partyId, instanceGuid, { action: action.name, buttonId });
+    },
+  });
+  const { handleClientActions, handleDataModelUpdate } = useHandleClientActions();
+
+  return {
+    mutation,
+    /**
+     * We wrap the mutation in a promise to make it possible to await both the mutation,
+     * and the side effects that are run in the onSuccess or onError callbacks.
+     */
```

Comment:
Are you sure you can't just use `mutation.mutateAsync()`? :thinking: It returns a promise, but you'll have to catch and handle any errors yourself (instead of checking `mutation.error`)

Reply:
Thats a great point! Yeah we can. I'm refactoring it, and also renaming the hook to `useHandleServerAction`  and the function returned as `handleServerAction`. That feels a little more consistent with the other hooks.